incompatibility/incompatibility-report/oas3-traverse.go

```diff
@@ -0,0 +1,160 @@
+// performs error checking for these requirements and casts into generic type
+func makeGenericIncompatibilityReportFunc(function interface{}) (func(interface{}) (*IncompatibilityReport, error), error) {
+	genGeneric, err := makeGenericOperation(function)
+	badType := genGeneric.outputType != reflect.TypeOf(&IncompatibilityReport{})
+	if err != nil || badType {
+		return nil, err
+	} else {
+		return func(i interface{}) (*IncompatibilityReport, error) {
+			res, err := genGeneric.op(i)
+			return res.(*IncompatibilityReport), err
+		}, nil
+	}
+}
+
+// ComponentType - component type in OpenAPIDoc, i.e. Parameters, Components, ...
+// ParentTraverse - lambda to tranverse from parent to component of type ComponentType
+// OperationOnComponent - lambda which looks at current layer for incompatibilities
+// ChildPaths - paths to check for deeper incompatibilities
+type PathOperation struct {
+	ComponentType        reflect.Type
+	ParentTraverse       GenericOperation
+	OperationOnComponent func(currentComponent interface{}) (*IncompatibilityReport, error)
+	ChildPaths           []*PathOperation
+}
+
+// Merge Two Path Operations under the conditions:
+// 1) Both Operations happen at same component
+//   - Thus compose a new OperationOnComponent
+// 2) p2's parent type is the same as p1's component type
+//   - add p2 to p1 child paths
+func (p1 PathOperation) compose(p2 PathOperation) (PathOperation, error) {
+	sameComponentType := p1.ComponentType == p2.ComponentType
+	sameTraversalType := p1.ParentTraverse.typeEquality(p2.ParentTraverse) == SameType
+	if sameComponentType && sameTraversalType {
+		// Assume operations take place along same component in path, thus merge
+		var composedComponentOperation = func(currentComponent interface{}) (*IncompatibilityReport, error) {
+			p1Report, err := p1.OperationOnComponent(currentComponent)
+			if err != nil {
+				return &IncompatibilityReport{}, err
+			}
+			p2Report, err := p2.OperationOnComponent(currentComponent)
+			if err != nil {
+				return &IncompatibilityReport{}, err
+			}
+			return &IncompatibilityReport{
+				Incompatibilities: append(p1Report.GetIncompatibilities(), p2Report.GetIncompatibilities()...)}, nil
+		}
+		return PathOperation{
+			ComponentType:        p1.ComponentType,
+			ParentTraverse:       p1.ParentTraverse,
+			OperationOnComponent: composedComponentOperation,
+			ChildPaths:           append(p1.ChildPaths, p2.ChildPaths...),
+		}, nil
+	}
+	//Try adding p2 to p1 children
+	p2parentType := p2.ParentTraverse.inputType
+	parent2childtraverse := p1.ComponentType == p2parentType
+	if parent2childtraverse { //
+		p1.ChildPaths = append(p1.ChildPaths, &p2)
+		return p1, nil
+	}
+	return PathOperation{}, TraverseError{"Could not merge Path Operations"}
+}
+
+// compile Path Operations, return incompatibility report from paths
+func (p PathOperation) compile(currentLevel interface{}) (*IncompatibilityReport, error) {
```

Comment:
This to me reads more like a "Apply" than a compile, because you're doing the path operation on a spec.